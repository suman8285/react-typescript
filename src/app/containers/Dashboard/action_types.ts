export const USER_BALANCE = 'USER_BALANCE';
export const USER_BALANCE_ERROR = 'USER_BALANCE_ERROR';
export const USER_BALANCE_SUCCESS = 'USER_BALANCE_SUCCESS';
