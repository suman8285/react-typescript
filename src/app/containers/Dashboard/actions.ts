import {
  USER_BALANCE,
  USER_BALANCE_ERROR,
  USER_BALANCE_SUCCESS,
} from './action_types';
import { IUserBalance, IDashboardState } from './types';

interface IAction {
  type: string;
  payload?: string | number;
}

export const getBalance = (): IAction => {
  return {
    type: USER_BALANCE,
  };
};

export const getBalanceSuccess = (payload: number): IAction => {
  return {
    type: USER_BALANCE_SUCCESS,
    payload,
  };
};

export const getBalanceError = (payload: string): IAction => {
  return {
    type: USER_BALANCE_ERROR,
    payload,
  };
};
