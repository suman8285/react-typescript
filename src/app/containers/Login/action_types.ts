export const AUTH_INITIALIZE = 'AUTH_INITIALIZE';
export const AUTH_LOGGING_IN = 'AUTH_LOGGING_IN';
export const AUTH_LOGGED_IN = 'AUTH_LOGGED_IN';
export const AUTH_ERROR = 'AUTH_ERROR';
export const AUTH_LOGOUT = 'AUTH_LOGOUT';
