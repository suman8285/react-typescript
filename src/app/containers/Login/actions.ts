import {
  AUTH_INITIALIZE,
  AUTH_LOGGING_IN,
  AUTH_LOGGED_IN,
  AUTH_ERROR,
  AUTH_LOGOUT,
} from './action_types';
import { IAuthCredentials } from './types';

interface IAction {
  type: string;
  payload?: string | IToken | IAuthCredentials;
}

interface IToken {
  [key: string]: string;
}

export const authInitialize = (payload: IToken): IAction => {
  return {
    type: AUTH_INITIALIZE,
    payload,
  };
};

export const authLoggingIn = (payload: IAuthCredentials): IAction => {
  return {
    type: AUTH_LOGGING_IN,
    payload,
  };
};

export const authLoggedIn = (payload: IToken): IAction => {
  return {
    type: AUTH_LOGGED_IN,
    payload,
  };
};

export const authError = (payload: string): IAction => {
  console.log('in auth error login');

  return {
    type: AUTH_ERROR,
    payload,
  };
};

export const authLogout = (): IAction => {
  return {
    type: AUTH_LOGOUT,
  };
};
