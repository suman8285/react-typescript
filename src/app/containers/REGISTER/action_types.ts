export const USER_REGISTRATION = 'USER_REGISTRATION';
export const USER_REGISTRATION_ERROR = 'USER_REGISTRATION_ERROR';
export const USER_REGISTRATION_SUCCESS = 'USER_REGISTRATION_SUCCESS';
