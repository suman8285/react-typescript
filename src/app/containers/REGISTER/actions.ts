import {
  USER_REGISTRATION,
  USER_REGISTRATION_ERROR,
  USER_REGISTRATION_SUCCESS,
} from './action_types';
import { IRegisterUser } from './types';

interface IAction {
  type: string;
  payload?: string | IRegisterUser;
}

export const userRegistring = (payload: IRegisterUser): IAction => {
  console.log('In this');

  return {
    type: USER_REGISTRATION,
    payload,
  };
};

export const userRegistrationSuccess = (): IAction => {
  console.log('in action2');

  return {
    type: USER_REGISTRATION_SUCCESS,
  };
};

export const userRegistrationError = (payload: string): IAction => {
  console.log('in action2', payload);

  return {
    type: USER_REGISTRATION_ERROR,
    payload,
  };
};
