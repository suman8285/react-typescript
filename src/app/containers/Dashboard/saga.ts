import { put, takeLatest, call } from 'redux-saga/effects';
import { PayloadAction } from '@reduxjs/toolkit';
import { request } from 'utils/request';
import { WalletEndpoint } from 'app/components/common/Constants';
import { IUserBalance } from './types';
import { USER_BALANCE } from './action_types';
import { getBalance, getBalanceSuccess, getBalanceError } from './actions';

interface IResponse {
  data: {
    walletBalance: number;
  };
}

export function* balanceEnquire(): Generator {
  try {
    const { data } = (yield call(request, {
      url: WalletEndpoint,
      method: 'get',
    })) as IResponse;

    yield put(getBalanceSuccess(data.walletBalance));
  } catch (err) {
    const errorMessage =
      err.response.status === 404
        ? 'User Not Found'
        : err.response.data.message;
    yield put(getBalanceError(errorMessage));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* balanceSaga(): Generator {
  // Watches for loadRepos actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(USER_BALANCE, balanceEnquire);
}
