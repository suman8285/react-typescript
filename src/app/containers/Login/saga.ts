import { put, takeLatest, call } from 'redux-saga/effects';
import { PayloadAction } from '@reduxjs/toolkit';
import { request } from 'utils/request';
import { LoginEndpoint } from 'app/components/common/Constants';
import { IAuthCredentials, ILoginUser } from './types';
import { AUTH_LOGGING_IN } from './action_types';
import { authLoggedIn, authError } from './actions';

interface IResponse {
  data: ILoginUser;
}

export function* getToken(action: PayloadAction<IAuthCredentials>): Generator {
  try {
    const { username, password } = action.payload;
    const { data } = (yield call(request, {
      url: LoginEndpoint,
      method: 'post',
      data: {
        userId: username,
        password,
      },
    })) as IResponse;

    yield put(authLoggedIn({ token: data.token, username: data.username }));
  } catch (err) {
    const errorMessage =
      err.response.status === 404
        ? 'User Not Found'
        : err.response.data.message;
    yield put(authError(errorMessage));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* authSaga(): Generator {
  // Watches for loadRepos actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(AUTH_LOGGING_IN, getToken);
}
