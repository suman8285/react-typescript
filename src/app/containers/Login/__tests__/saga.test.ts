import { put, takeLatest } from 'redux-saga/effects';
import { authSaga, getToken } from '../saga';
import { AUTH_LOGGING_IN } from '../action_types';
import { authLoggedIn, authError } from '../actions';

describe('getToken Saga', () => {
  let token: string | null;
  let getTokenIterator: ReturnType<typeof getToken>;
  let authInitializePutDescriptor;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getTokenIterator = getToken({
      type: AUTH_LOGGING_IN,
      payload: {
        username: 'Nila',
        password: '1234456',
      },
    });
    authInitializePutDescriptor = getTokenIterator.next().value;

    expect(authInitializePutDescriptor).toMatchSnapshot();
  });

  it('should dispatch the authLoggedIn action if it requests the data successfully', () => {
    token =
      'dhsfhdijskljdlklkhgskljjgklsd.dbvjksbsjhijrgokodl.bchdfgakeufhaufhio';
    const response = {
      data: {
        token,
      },
    };

    const putDescriptor = getTokenIterator.next(response).value;
    expect(putDescriptor).toEqual(put(authLoggedIn({ token })));
  });

  it('should dispatch user not found', () => {
    const putDescriptor = getTokenIterator.throw({ response: { status: 404 } })
      .value;
    expect(putDescriptor).toEqual(put(authError('User Not Found')));
  });

  it('should dispatch unknown error', () => {
    const putDescriptor = getTokenIterator.throw({
      response: {
        status: 401,
        data: {
          message: 'Invalid password',
        },
      },
    }).value;
    expect(putDescriptor).toEqual(put(authError('Invalid password')));
  });
});

describe('authSaga Saga', () => {
  const authIterator = authSaga();
  it('should start task to watch for auth login action', () => {
    const takeLatestDescriptor = authIterator.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(AUTH_LOGGING_IN, getToken));
  });
});
