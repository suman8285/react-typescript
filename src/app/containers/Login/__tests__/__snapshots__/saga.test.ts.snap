// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getToken Saga should dispatch the authLoggedIn action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "data": Object {
          "password": "1234456",
          "userId": "Nila",
        },
        "method": "post",
        "url": "/login/user",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getToken Saga should dispatch unknown error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "data": Object {
          "password": "1234456",
          "userId": "Nila",
        },
        "method": "post",
        "url": "/login/user",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getToken Saga should dispatch user not found 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "data": Object {
          "password": "1234456",
          "userId": "Nila",
        },
        "method": "post",
        "url": "/login/user",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
